sdks/java/core/src/main/java/org/apache/beam/sdk/io/BigQueryIO.java

```diff
@@ -569,10 +613,353 @@ public boolean getValidate() {
+
+      private static List<String> executeExtract(
+          String jobId, TableReference table, String tempLocation, JobService jobService)
+              throws InterruptedException, IOException {
+        String uuid = UUID.randomUUID().toString();
+        String destinationUri =
+            String.format("%s/%s/%s", tempLocation, uuid, "*.avro");
+
+        JobConfigurationExtract extract = new JobConfigurationExtract();
+        extract.setSourceTable(table);
+        extract.setDestinationFormat("AVRO");
+        extract.setDestinationUris(ImmutableList.of(destinationUri));
+
+        jobService.startExtractJob(jobId, extract);
+        Job job =
+            jobService.pollJobStatus(table.getProjectId(), jobId, MAX_JOB_STATUS_POLL_RETRIES);
+        if (parseStatus(job) != Status.SUCCEEDED) {
+          throw new IOException("Extract job failed: " + jobId);
+        }
+
+        JobStatistics jobStats = job.getStatistics();
+        LOG.info("JobStats: " + jobStats);
+        List<Long> counts = jobStats.getExtract().getDestinationUriFileCounts();
+        checkState(counts.size() == 1, "Expect one count for the destination uri.");
+        long filesCount = counts.get(0);
+        List<String> tempFiles = Lists.newArrayList();
+        for (int i = 0; i < filesCount; ++i) {
+          tempFiles.add(
+              String.format("%s/%s/%012d%s", tempLocation, uuid, i, ".avro"));
+        }
+        return ImmutableList.copyOf(tempFiles);
+      }
+
+      private List<BoundedSource<TableRow>> createSources(
+          List<String> files, TableSchema tableSchema) throws IOException, InterruptedException {
+        final String jsonSchema = JSON_FACTORY.toString(tableSchema);
+
+        SerializableFunction<GenericRecord, TableRow> function =
+            new SerializableFunction<GenericRecord, TableRow>() {
+              @Override
+              public TableRow apply(GenericRecord input) {
+                try {
+                  return AvroUtils.convertGenericRecordToTableRow(
+                      input, JSON_FACTORY.fromString(jsonSchema, TableSchema.class));
+                } catch (IOException e) {
+                  throw Throwables.propagate(e);
+                }
+              }};
+
+        List<BoundedSource<TableRow>> avroSources = Lists.newArrayList();
+        BackOff backoff = new AttemptBoundedExponentialBackOff(
+            MAX_FILES_VERIFY_ATTEMPTS, INITIAL_FILES_VERIFY_BACKOFF_MILLIS);
+        for (String fileName : files) {
+          while (BackOffUtils.next(Sleeper.DEFAULT, backoff)) {
+            if (IOChannelUtils.getFactory(fileName).getSizeBytes(fileName) != -1) {
+              break;
+            }
+          }
+          avroSources.add(new TransformingSource<>(
+              AvroSource.from(fileName), function, getDefaultOutputCoder()));
+        }
+        return ImmutableList.copyOf(avroSources);
+      }
+    }
+
+    private static class TransformingSource<T, V> extends BoundedSource<V> {
```

Comment:
Oh, I see, this is generic. I wonder if this would be useful more broadly.

If so, consider extracting it to its own file (feel free to put it in utils until we know what to do with it) and test it. If not, probably should make it TableRow-specific